crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1908 @@
+        return 0;
+    encode_pubkey(out, key);
+    return 1;
+}
+
+/* Serialise an ML-KEM private key */
+int ossl_ml_kem_encode_private_key(uint8_t *out, size_t len,
+                                   const ML_KEM_KEY *key)
+{
+    if (!ossl_ml_kem_have_prvkey(key)
+        || len != key->vinfo->prvkey_bytes)
+        return 0;
+    encode_prvkey(out, key);
+    return 1;
+}
+
+/* Parse input as a public key */
+int ossl_ml_kem_parse_public_key(const uint8_t *in, size_t len, ML_KEM_KEY *key)
+{
+    EVP_MD_CTX *mdctx = NULL;
+    const ML_KEM_VINFO *vinfo;
+    int ret = 0;
+
+    /* Keys with key material are immutable */
+    if (key == NULL || ossl_ml_kem_have_pubkey(key))
+        return 0;
+    vinfo = key->vinfo;
+
+    if (len != vinfo->pubkey_bytes
+        || (mdctx = EVP_MD_CTX_new()) == NULL)
+        return 0;
+
+    if (add_storage(OPENSSL_malloc(vinfo->puballoc), 0, key))
+        ret = parse_pubkey(in, mdctx, key);
+
+    if (!ret)
+        free_storage(key);
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+/* Parse input as a new private key  */
+int ossl_ml_kem_parse_private_key(const uint8_t *in, size_t len,
+                                  ML_KEM_KEY *key)
+{
+    EVP_MD_CTX *mdctx = NULL;
+    const ML_KEM_VINFO *vinfo;
+    int ret = 0;
+
+    /* Keys with key material are immutable */
+    if (key == NULL || ossl_ml_kem_have_pubkey(key))
+        return 0;
+    vinfo = key->vinfo;
+
+    if (len != vinfo->prvkey_bytes
+        || (mdctx = EVP_MD_CTX_new()) == NULL)
+        return 0;
+
+    if (add_storage(OPENSSL_malloc(vinfo->prvalloc), 1, key))
+        ret = parse_prvkey(in, mdctx, key);
+
+    if (!ret)
+        free_storage(key);
+    EVP_MD_CTX_free(mdctx);
+    return ret;
+}
+
+/*
+ * Generate a new keypair from a given seed, giving a deterministic result for
+ * running tests.  The caller can elect to not collect the encoded public key.
+ */
+int ossl_ml_kem_genkey_seed(const uint8_t *seed, size_t seedlen,
+                            uint8_t *pubenc, size_t publen,
+                            ML_KEM_KEY *key)
+{
+    EVP_MD_CTX *mdctx = NULL;
+    const ML_KEM_VINFO *vinfo;
+    int ret = 0;
+
+    if (key == NULL || ossl_ml_kem_have_pubkey(key))
+        return 0;
+    vinfo = key->vinfo;
+
+    if (seed == NULL || seedlen != ML_KEM_SEED_BYTES
+        || (pubenc != NULL && publen != vinfo->pubkey_bytes)
+        || (mdctx = EVP_MD_CTX_new()) == NULL)
+        return 0;
+
+    if (add_storage(OPENSSL_malloc(vinfo->prvalloc), 1, key)) {
+        const uint8_t *d = seed;
+        const uint8_t *z = seed + ML_KEM_RANDOM_BYTES;
+
+        /*-
+         * This avoids the need to handle allocation failures for two (max 2KB
+         * each) vectors and (if the caller does not want the public key) an
+         * encoded public key (max 1568 bytes), that are never retained on
+         * return from this function.
+         * We stack-allocate these.
+         */
+#       define case_genkey_seed(bits) \
```

Comment:
We'd not normally indent the define like this.  Similar for the undef below.

Reply:
We also rarely put defines in the middle of loop blocks. I think this is warranted here to maintain indentation for readability.